File: Binary_Tree/513.Find_Bottom_Left_Tree_Value/C++/513.Find_Bottom_Left_Tree_Value.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int findBottomLeftValue(TreeNode * root) 
    {
        //=================================//
        // Input type:                     //
        //   - TreeNode * (root tree-node) //
        // Return type:                    //
        //   - int (result node value)     //
        //=================================//

        /*Initialize*/
        ///// Record queue /////
        queue<TreeNode * > record_queue;

        //// Result node value /////
        int res_val = INT_MAX;


        /*Breath-first search based loop traversal*/
        record_queue.emplace(root); //Update, Record

        while (! record_queue.empty()) //Cycled-times
        {
            bool record_flag = false; //Record flag

            for (int queue_idx = record_queue.size(); (queue_idx > 0); (queue_idx--)) //Queue-size
            {
                TreeNode * record_root = record_queue.front(); record_queue.pop(); //Keep updating, popped

                ///// Check if the current flag matched conditions or not /////
                if (! record_flag) { res_val = (record_root->val); record_flag |= true; } //Keep updating, overwriting
                else { ; }

                ///// Check if the next root left-half node existed or not /////
                if ((record_root->left) != nullptr) { record_queue.emplace((record_root->left)); } //Keep updating, recording
                else { ; }

                ///// Check if the next root right-half node existed or not /////
                if ((record_root->right) != nullptr) { record_queue.emplace((record_root->right)); } //Keep updating, recording
                else { ; }

            } //Queue-size

        } //Cycled-times

        return res_val;
    }
};

File: Binary_Tree/513.Find_Bottom_Left_Tree_Value/Python/513.Find_Bottom_Left_Tree_Value.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def findBottomLeftValue(self, root):
        """
        :type root: Optional[TreeNode]
        :rtype: int
        """
        #============================================#
        # Breath-first search based traversal method #
        #============================================#

        ############
        #Initialize
        ##### Record queue ######
        record_queue = [root]

        ##### Result node value #####
        res_val = float("inf")


        ##########################################
        #Breath-first search based loop traversal
        while (record_queue):
            len_queue = len(record_queue) #Length of queue

            record_flag = False #Record flag

            for queue_idx in range(len_queue):
                record_root = record_queue.pop(0) #Record root

                ##### Check if the current flag matched conditions or not #####
                if (not record_flag):
                    res_val = record_root.val #Keep updating/overwriting

                    record_flag |= True #Keep updating/overwriting
                else:
                    pass

                ##### Check if the next root left-half node existed or not #####
                if (record_root.left):
                    record_queue.append(record_root.left) #Keep updating/recording
                else:
                    pass

                ##### Check if the next root right-half node existed or not #####
                if (record_root.right):
                    record_queue.append(record_root.right) #Keep updating/recording
                else:
                    pass

        return res_val
